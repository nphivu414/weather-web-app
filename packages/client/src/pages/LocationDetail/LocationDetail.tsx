import * as React from 'react';
import AppBar from 'components/AppBar';
import NavigationContext from 'navigation';
import { fetchLocationDetail } from 'services';
import { useQuery } from 'react-query';
import TodayWeatherSection from './TodayWeatherSection';
import Spinner from 'components/Spinner';
import DailyWeatherSection from './DailyWeatherSection';
import { Place, ServiceError, Weather } from 'types';

export type LocationDetailProps = {
  id: number | null;
};

export const renderTodayWeatherSection = (todayWeather: Weather) => {
  if (!todayWeather) {
    return null;
  }

  const { min_temp, max_temp, the_temp, weather_state_name } = todayWeather;
  return (
    <TodayWeatherSection
      minTemp={min_temp}
      maxTemp={max_temp}
      currentTemp={the_temp}
      weatherStateName={weather_state_name}
    />
  );
};

export const LocationDetail: React.FC<LocationDetailProps> = ({ id }) => {
  const { goBack, title } = React.useContext(NavigationContext);
  const { data, isLoading, error, isError } = useQuery<Place | null, ServiceError>(
    ['locationDetail', id],
    fetchLocationDetail({
      id,
    }),
    {
      enabled: id !== null,
    },
  );
  const weatherList = data?.consolidated_weather || [];
  const todayWeather = weatherList[0];

  return (
    <div>
      <AppBar title={title} hasBackButton onBack={goBack} />
      <div className="p-3 bg-white">
        {isError && !data && (
          <div className="alert alert-danger text-center" role="alert">
            <label>{error?.message}</label>
          </div>
        )}
        <Spinner isSpinning={isLoading} />
        {renderTodayWeatherSection(todayWeather)}
        <DailyWeatherSection weatherList={weatherList} />
      </div>
    </div>
  );
};

export default LocationDetail;
