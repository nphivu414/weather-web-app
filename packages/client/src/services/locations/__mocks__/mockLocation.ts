import { Place, Weather } from 'types';
export const mockLocations: Place[] = [
  {
    title: 'Helsinki',
    location_type: 'City',
    woeid: 144246,
    latt_long: '13.115,-0.411'
  }
]

export const mockLocationDetail: Place = {
  consolidated_weather: [
    {
      id: '4811453371514880',
      weather_state_name: 'Light Rain',
      weather_state_abbr: 'lr',
      wind_direction_compass: 'S',
      created: '2021-01-05T03:20:16.354553Z',
      applicable_date: '2021-01-04',
      min_temp: 10.335,
      max_temp: 14.04,
      the_temp: 12.84,
      wind_speed: 6.662041035769771,
      wind_direction: 179.13574633334716,
      air_pressure: 1022,
      humidity: 87,
      visibility: 15.07757197963891,
      predictability: 75
    },
  ],
  title: 'San Francisco',
  location_type: 'City',
  woeid: 2487956,
  latt_long: '37.777119, -122.41964',
}

export const mockLocationWeather: Weather = {
  id: '5219640084004864',
  weather_state_name: 'Heavy Cloud',
  weather_state_abbr: 'hc',
  wind_direction_compass: 'NE',
  created: '2021-01-04T09:22:58.583752Z',
  applicable_date: '2021-01-04',
  min_temp: 1.18,
  max_temp: 8.535,
  the_temp: 7.65,
  wind_speed: 8.627662748028845,
  wind_direction: 49.82635538871968,
  air_pressure: 1021.5,
  humidity: 76,
  visibility: 9.605466575200827,
  predictability: 71,
};

export const mockLocationWeatherList: Weather[] = [
  {
    id: '5219640084004864',
    weather_state_name: 'Heavy Cloud',
    weather_state_abbr: 'hc',
    wind_direction_compass: 'NE',
    created: '2021-01-04T09:22:58.583752Z',
    applicable_date: '2021-01-04',
    min_temp: 1.18,
    max_temp: 8.535,
    the_temp: 7.65,
    wind_speed: 8.627662748028845,
    wind_direction: 49.82635538871968,
    air_pressure: 1021.5,
    humidity: 76,
    visibility: 9.605466575200827,
    predictability: 71,
  },
  {
    id: '4892388808982528',
    weather_state_name: 'Heavy Rain',
    weather_state_abbr: 'hr',
    wind_direction_compass: 'SW',
    created: '2021-01-04T12:20:17.765280Z',
    applicable_date: '2021-01-04',
    min_temp: 9.504999999999999,
    max_temp: 12.379999999999999,
    the_temp: 13.325,
    wind_speed: 11.596992554993506,
    wind_direction: 216.04166545749246,
    air_pressure: 1018.5,
    humidity: 92,
    visibility: 8.728090451761712,
    predictability: 77,
  }
]
