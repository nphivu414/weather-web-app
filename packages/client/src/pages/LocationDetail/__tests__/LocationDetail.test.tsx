import * as React from 'react';
import { shallow } from 'enzyme';

import LocationDetail, { LocationDetailProps, renderTodayWeatherSection } from '../LocationDetail';
import { QueryClient, QueryClientProvider } from 'react-query';
import { mockLocationWeather } from 'services/locations/__mocks__/mockLocation';

const props: LocationDetailProps = {
  id: 1,
};

const queryClient = new QueryClient();

describe('LocationDetail', () => {
  it('should match a snapshot', () => {
    expect(
      shallow(
        <QueryClientProvider client={queryClient}>
          <LocationDetail {...props} />
        </QueryClientProvider>,
      ),
    ).toMatchSnapshot();
  });

  describe('renderTodayWeatherSection', () => {
    it('should match a snapshot', () => {
      expect(shallow(renderTodayWeatherSection(mockLocationWeather))).toMatchSnapshot();
    });
  });
});
