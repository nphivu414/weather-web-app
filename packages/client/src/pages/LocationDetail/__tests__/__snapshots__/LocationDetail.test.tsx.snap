// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationDetail renderTodayWeatherSection should match a snapshot 1`] = `ShallowWrapper {}`;

exports[`LocationDetail should match a snapshot 1`] = `ShallowWrapper {}`;
